### PR TITLE
Add a setting to disable TLS for sending email

### LINKED ISSUE #8046: Unable to configure SMTP to use TLS without STARTTLS
### Description

Hi. I'm the sysadmin operating converser.eu homeserver. It seems that my users can't register email in their accounts. New users can't register when they put an email.

When I try to add an email, there is a POST request to `https://converser.eu/_matrix/client/r0/account/3pid/email/requestToken` but it timeouts. On app.element.io, I've got the error `request failed: ETIMEDOUT` .

In synapse configuration, the identity server is defined. The mailserver is configured too, and the connection to it is ok.

I've got nginx as a proxy. I've got a timeout : 
`2020/08/06 15:32:09 [error] 1067425#1067425: *1948 upstream timed out (110: Connection timed out) while reading response header from upstream, client: 2a0e:xxxxxx, server: converser.eu, request: "POST /_matrix/client/r0/account/3pid/email/requestToken HTTP/2.0", upstream: "http://127.0.0.1:8008/_matrix/client/r0/account/3pid/email/requestToken", host: "converser.eu"`
But no logs related to this in homeserver.log.

Am I missing something ?

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**: converser.eu

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: 1.17.0

- **Install method**: debian package
<!-- examples: package manager/git clone/pip  -->

- **Platform**: debian buster + backports
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

### PR DESCRIPTION
**Pull request by richvdh**
This is mostly useful in case the server offers TLS, but doesn't present a valid certificate.

Fixes #8046.

### CODE REVIEW

File: synapse/handlers/send_email.py [hunk 1]
```diff
@@ -26,19 +31,75 @@
...
+    smtpport: int,
+    from_addr: str,
+    to_addr: str,
+    msg_bytes: bytes,
+    username: Optional[bytes] = None,
+    password: Optional[bytes] = None,
+    require_auth: bool = False,
+    require_tls: bool = False,
+    tls_hostname: Optional[str] = None,
+):
```
Comment (erikjohnston) [inline synapse/handlers/send_email.py:46]:
```suggestion
) -> None:
```

It doesn't look like we define or use the return value?

File: synapse/handlers/send_email.py [hunk 2]
```diff
@@ -26,19 +31,75 @@
...
+        heloFallback=True,
+        requireAuthentication=require_auth,
+        requireTransportSecurity=require_tls,
+        hostname=tls_hostname,
+    )
+
+    # the IReactorTCP interface claims host has to be a bytes, which seems to be wrong
+    reactor.connectTCP(smtphost, smtpport, factory, timeout=30, bindAddress=None)  # type: ignore[arg-type]
+
+    return await make_deferred_yieldable(d)
```
Comment (erikjohnston) [inline synapse/handlers/send_email.py:82]:
```suggestion
    await make_deferred_yieldable(d)
```

File: synapse/handlers/send_email.py [hunk 3]
```diff
@@ -26,19 +31,75 @@
...
+        msg_bytes: Message content
+        username: username to authenticate with, if auth is enabled
+        password: password to give when authenticating
+        require_auth: if auth is not offered, fail the request
+        require_tls: if TLS is not offered, fail the reqest
+        tls_hostname: TLS hostname to check for. None to disable TLS.
+    """
+    msg = BytesIO(msg_bytes)
+
+    d: "Deferred[object]" = Deferred()
```
Comment (erikjohnston) [inline synapse/handlers/send_email.py:64]: Why `object` ooi?

Comment (richvdh) [reply synapse/handlers/send_email.py:64]: ugh, I flipflopped on this a bit.

It turns out that the deferred is callbacked with a tuple of someting like `(number of messages, [(result1_tuple), (result2_tuple)])`. I didn't really want to get into defining a type for that, so just went with `object`.

Comment (erikjohnston) [reply synapse/handlers/send_email.py:64]: Fair, I was mostly just expecting `Any` there, but :shrug:

**Review by erikjohnston — APPROVED** [review]

**Pull request merged**